akka-stream/src/main/scala/akka/stream/impl/fusing/Fusing.scala

```diff
@@ -0,0 +1,362 @@
+    var it = group.iterator
+    val ups = struct.upstreams
+    val downs = struct.downstreams
+    val outGroup = struct.outGroup
+    while (it.hasNext) it.next() match {
+      case copy @ CopiedModule(shape, attr, gsm: GraphStageModule) ⇒
+        stages(pos) = gsm.stage
+        matValIDs(pos) = copy
+        attributes(pos) = attr
+
+        shape.inlets.iterator.zip(gsm.shape.inlets.iterator).foreach {
+          case (in, orig) ⇒
+            val out = ups.get(in)
+            val internal = (out != null) && (outGroup.get(out) eq group)
+            if (internal) {
+              ups.remove(in)
+              downs.remove(out)
+              outConns.put(out, insB2.size)
+              insB2.add(orig)
+              inOwnersB2.add(pos)
+            } else {
+              insB1.add(orig)
+              inOwnersB1.add(pos)
+              inlets.add(in)
+            }
+        }
+
+        pos += 1
+    }
+
+    val outsB2 = new Array[Outlet[_]](insB2.size)
+    val outOwnersB2 = new Array[Int](insB2.size)
+
+    pos = 0
+    it = group.iterator
+    while (it.hasNext) it.next() match {
+      case CopiedModule(shape, _, gsm: GraphStageModule) ⇒
+        shape.outlets.iterator.zip(gsm.shape.outlets.iterator).foreach {
+          case (out, orig) ⇒
+            if (outConns.containsKey(out)) {
+              val idx = outConns.remove(out)
+              outsB2(idx) = orig
+              outOwnersB2(idx) = pos
+            } else {
+              outsB3.add(orig)
+              outOwnersB3.add(pos)
+              outlets.add(out)
+            }
+        }
+        pos += 1
+    }
+
+    val shape = AmorphousShape(inlets.asScala.to[immutable.Seq], outlets.asScala.to[immutable.Seq])
+
+    val connStart = insB1.size
+    val conns = insB2.size
+    val outStart = connStart + conns
+    val size = outStart + outsB3.size
+
+    val ins = new Array[Inlet[_]](size)
+    copyToArray(insB2.iterator, ins, copyToArray(insB1.iterator, ins, 0))
+
+    val inOwners = new Array[Int](size)
+    Arrays.fill(inOwners, copyToArray(inOwnersB2.iterator, inOwners, copyToArray(inOwnersB1.iterator, inOwners, 0)), size, -1)
+
+    val outs = new Array[Outlet[_]](size)
+    System.arraycopy(outsB2, 0, outs, connStart, conns)
+    copyToArray(outsB3.iterator, outs, outStart)
+
+    val outOwners = new Array[Int](size)
+    Arrays.fill(outOwners, 0, connStart, -1)
+    System.arraycopy(outOwnersB2, 0, outOwners, connStart, conns)
+    copyToArray(outOwnersB3.iterator, outOwners, outStart)
+
+    // FIXME attributes should contain some naming info and async boundary where needed
+    GraphModule(new GraphInterpreter.GraphAssembly(stages, attributes, ins, inOwners, outs, outOwners), shape, Attributes.none, matValIDs)
+  }
+
+  @tailrec private def copyToArray[T](it: ju.Iterator[T], array: Array[T], idx: Int): Int =
+    if (it.hasNext) {
+      array(idx) = it.next()
+      copyToArray(it, array, idx + 1)
+    } else idx
+
+  private def descend(m: Module,
```

Comment:
I would pretty much apprieciate a large internal doc block just like in the GraphInterpreter file explaining the overall phases, data structures and principles used. Then with some smaller comment blocks on the actual pieces. This can be a hell to maintain otherwise.